### PR TITLE
WFP-1550: Implement graph calls

### PR DESCRIPTION
**Pull request by sp-watson**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/DatabaseWriteService.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.justice.digital.hmpps.hmppsstafflookupservice.service
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppsstafflookupservice.client.MicrosoftADUser
+
+@Service
+class DatabaseWriteService {
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/DatabaseWriteService.kt:7]: This service will allow us to develop of the database layer separately.

The names will probably need changing.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/IndexingService.kt [hunk 2]
```diff
@@ -5,12 +5,16 @@ import uk.gov.justice.digital.hmpps.hmppsstafflookupservice.client.MicrosoftGrap
...
-    exampleData.collect {
-      println("DATA: $it")
+    var exampleData = microsoftGraphClient.getUsersPage(null)
+    databaseWriteService.writeData(exampleData.value)
+    while (exampleData.nextLink != null) {
+      val skipToken = exampleData.nextLink!!.substringAfter("skiptoken=")
+      exampleData = microsoftGraphClient.getUsersPage(skipToken)
+      databaseWriteService.writeData(exampleData.value)
     }
     return "DONE"
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/IndexingService.kt:19]: We will review this response later, when we make the call async

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/integration/indexing/FullReindex.kt [hunk 3]
```diff
@@ -16,4 +24,16 @@ class FullReindex : IntegrationTestBase() {
...
+  @Test
+  fun `submits the graph data for saving`() {
+    multiplePageGraphResponse()
+    webTestClient.post()
+      .uri("/admin/refresh-staffs")
+      .headers(setAuthorisation(roles = listOf("ROLE_TEST")))
+      .exchange()
+      .expectStatus()
+      .isOk
+    verify(databaseWriteService, times(2)).writeData(any())
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/integration/indexing/FullReindex.kt:37]: This expectation will be replaced/removed when we implement the DB calls.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/IndexingService.kt [hunk 4]
```diff
@@ -5,12 +5,16 @@ import uk.gov.justice.digital.hmpps.hmppsstafflookupservice.client.MicrosoftGrap
...
+  val databaseWriteService: DatabaseWriteService
 ) {
   suspend fun indexAll(): String {
-    val exampleData = microsoftGraphClient.getUsersWithoutPagination()
-    exampleData.collect {
-      println("DATA: $it")
+    var graphResponse = microsoftGraphClient.getUsersPage(null)
+    databaseWriteService.writeData(graphResponse.value)
+    while (graphResponse.nextLink != null) {
+      val skipToken = graphResponse.nextLink!!.substringAfter("skiptoken=")
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/IndexingService.kt:15]: is there ever a time when there isn't a skiptoken in the nextLink? Is it always at the end?

Comment (sp-watson) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/IndexingService.kt:15]: The response for the last page does not contain a nextLink (I tested that). If there isn't a nextLink then it won't get here..

**Review by carlov20 — APPROVED** [review]

**Pull request merged**